tests/src/app.module.ts

```diff
@@ -7,15 +7,44 @@ import { ConfigService } from '../../lib/config.service';
 import databaseConfig from './database.config';
 import nestedDatabaseConfig from './nested-database.config';
 
+type Config = {
+  database: ConfigType<typeof databaseConfig> & {
+    driver: ConfigType<typeof nestedDatabaseConfig>
+  };
+};
+
 @Module({})
 export class AppModule {
   constructor(
     private readonly configService: ConfigService,
+    // The following is the same object as above but narrowing its types
+    private readonly configServiceNarrowed: ConfigService<Config, true>,
     @Optional()
     @Inject(databaseConfig.KEY)
     private readonly dbConfig: ConfigType<typeof databaseConfig>,
   ) {}
 
+  /**
+   * This method is not meant to be used anywhere! It just here for testing
+   * types defintions while runnig test suites (in some sort).
+   * If some typings doesn't follows the requirements, Jest will fail due to
+   * TypeScript errors.
+   */
+  private noop(): void {
```

Comment:
I'm not sure if this strategy is that bad tho. Since I don't want to add libs such as `tsd` or `dtslint` to check if my proposing feature for `ConfigService` it working as expected (and keep working in the future), I've added this method do to the type checking as a side-effect of running `npm run test:integration` :p

I could drop this commit if you want to.